File: sieve2.pl
# use strict;
use integer;
print "Perl\n";
$Expected = 15485863;
my $N = 16000000; my $FINDPRIME = 1000000; my $maxprime = 0;
my $count = 1; my $i = 3; my $k = 0; my $N_sqrt = sqrt($N);
my @b = (0..$N);
for ($i = 3; $i <= $N_sqrt; $i += 2) {
    if ( defined $b[$i] ) {
        $count += 1;
        for ($k = $i * $i; $k <= $N; $k += $i * 2) {
            undef $b[$k];
        }
    }
}
for ( ; $i <= $N; $i += 2 ) {
    if (defined $b[$i]) {
        $maxprime=$i;
        $count += 1;
    }
    $i=$N + 1 if ($count >= $FINDPRIME);
}
print "Total Primes: $count\n";
print "Max Prime: $maxprime\n";
print "ERROR: Largest Prime was incorrect. Should be: $Expected\n" if ($maxprime != $Expected);

File: sieve2.rb
puts 'Ruby'
Expected = 15485863
N,Findprime = 16000000, 1000000
maxprime,count,N_sqrt = 0, 1, Integer(Math::sqrt(N))
b = Array.new(N)
3.step(N_sqrt, 2) do |i|
  next if b[i]
  count += 1
  (i*i).step(N, i << 1) { |k| b[k] = true }
end
N_sqrt.succ.step(N, 2) do |i|
  next if b[i]
  maxprime = i
  count += 1
  break if count >= Findprime
end
puts "Total Primes  = #{count}"
puts "Max Prime = #{maxprime}"
puts "ERROR: Largest Prime was incorrect. Should be: #{Expected}" if maxprime != Expected
